core/domain/classifier_services.py

```diff
@@ -424,41 +375,240 @@ def delete_classifier_training_job(job_id):
+            exp_id, exp_version, state_name))
+    if state_training_jobs_mapping_model is None:
+        return None
+    job_id = state_training_jobs_mapping_model.algorithm_ids_to_job_ids[
+        algorithm_id]
+    return get_classifier_training_job_by_id(job_id)
+
+
+def get_state_training_jobs_mapping(exp_id, exp_version, state_name):
+    """Gets training job exploration mapping model for given exploration state
+    combination.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+        state_name: str. Name of the state for which training job mapping model
+            is to be retrieved.
+
+    Returns:
+        StateTrainingJobsMapping. A domain object containing exploration
+        mapping model information.
+    """
+    state_training_jobs_mapping_model = (
+        classifier_models.StateTrainingJobsMappingModel.get_model(
+            exp_id, exp_version, state_name))
+    if state_training_jobs_mapping_model:
+        return classifier_domain.StateTrainingJobsMapping(
+            state_training_jobs_mapping_model.exp_id,
+            state_training_jobs_mapping_model.exp_version,
+            state_training_jobs_mapping_model.state_name,
+            state_training_jobs_mapping_model.algorithm_ids_to_job_ids)
+
+
+def migrate_exploration_training_jobs(state_training_jobs_mapping):
+    """Migrate exploration training jobs to latest version of algorithm_id
+    and algorithm_version.
+
+    This function lazily migrates an older classifier training job and
+    trains new classifiers. Specifically, if training job exploration mapping of
+    an <exploration, version, state> triplet is missing job_id for some
+    algorithm_id, or if the job_id exists but it has been trained on a now
+    obsolete algorithm, we re-submit the jobs.
+
+    The function goes through existing training job exploration mapping and
+    identifies three different types of algorithm IDs.
+        1. algorithm_ids_to_upgrade: Those which exist but needs to be
+            upgraded a new algorithm (because existing one has been deprecated)
+            by re-submitting the training job.
+        2. algorithm_ids_to_add: Those which doesn't exist and needs to be added
+            by submitting a new training job.
+        3. algorithm_ids_to_remove: Those which needs to be removed since these
+            algorithms are no longer supported.
+
+    Once all three types of algorithm IDs are filtered, the function performs
+    specific tasks tailored to each of them. Its called lazy migration because
+    it happens only when there is a query to retrieve trained model for given
+    <exploration, version, state> and algorithm_id.
+
+    Args:
+        state_training_jobs_mapping: StateTrainingJobsMapping. Domain
+            object containing exploration to training job id mapping. This
+            mapping is used to figure out jobs that needs to be re-submitted /
```

Comment:
needs --> need

change "/" to a comma, and delete the space before it.

Reply:
Done